### PR TITLE
Support validating target reports through image stream

### PR DESCRIPTION
**Pull request by iche033**
Depends on https://github.com/ignitionrobotics/ign-gazebo/pull/1381

Targets are to be reported by labeling their location in a live video stream. This PR adds the following functionality:
* ign service to indicate a stream has started
* ign service for reporting the target location (image 2d pos) in the image stream
* validate the reported image pos contains a valid target

The GameLogic plugin does not actually process any incoming video images, e.g. no object recognition or image processing, etc but instead it uses the rendering engine to check if target objects are in the camera view (by checking target is within frustum and projecting ray to see if it hits the target)

To test, I find it easiest to make the UAV static first so it does not fall when we hover it over the target, e.g.

```diff
diff --git a/mbzirc_ign/models/mbzirc_quadrotor/model.sdf.erb b/mbzirc_ign/models/mbzirc_quadrotor/model.sdf.erb
index 838c1b4..51960c3 100644
--- a/mbzirc_ign/models/mbzirc_quadrotor/model.sdf.erb
+++ b/mbzirc_ign/models/mbzirc_quadrotor/model.sdf.erb
@@ -75,6 +75,7 @@ end
 <?xml version="1.0"?>
 <sdf version='1.9'>
   <model name="<%= $model_name%>">
+    <static>true</static>

```

Make sure to build and install the workspace.

1. Launch simple demo world:

    ```sh
    ros2 launch ros_ign_gazebo ign_gazebo.launch.py ign_args:="-v 4 -r simple_demo.sdf"
    ```
1. spawn quadrotor with downward looking camera:

    ```sh
    # slot3 is downward looking camera
    ros2 launch mbzirc_ign spawn.launch.py name:=quadrotor_1 world:=simple_demo model:=mbzirc_quadrotor type:=uav x:=-0 y:=0 z:=1.2 R:=0 P:=0 Y:=0 slot3:=mbzirc_hd_camera
    ```

1. Move the quadrotor above vessel:

    ```sh
    # adjust your x and y pos accordingly since vessel may move due to wave and wind
    ign service -s /world/simple_demo/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "quadrotor_1", position: {x: 25.0, y:25, z:80}'
    ```

1. Use ign service to indicate that stream is to be started

    ```sh
    # string args are: name of vehicle and sensor producing the video stream
    ign service -s /mbzirc/target/stream/start --reqtype ignition.msgs.StringMsg_V --reptype ignition.msgs.Boolean --timeout 300 --req 'data: ["quadrotor_1", "sensor_3"]'
    ```

1. Use ign service to report target vessel in stream

    ```sh
    # 640, 480 is center of image
    ign service -s /mbzirc/target/stream/report --reqtype ignition.msgs.StringMsg_V --reptype ignition.msgs.Boolean --timeout 300 --req 'data: ["vessel", "640", "480"]'
    ```

1. Open `/tmp/ign/mbzirc/logs/events.yml` to see that the vessel is successfully reported. You should see an event with data: `vessel_id_success`

1. Repeat the process for small and large target objects

    ```sh
    # adjust your x and y pos accordingly since vessel may move due to wave and wind
    # set low z pos so it can see the obj without hitting the mainsail part of the boat
    ign service -s /world/simple_demo/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "quadrotor_1", position: {x: 24.0, y:26.5, z:2}'

    ign service -s /mbzirc/target/stream/report --reqtype ignition.msgs.StringMsg_V --reptype ignition.msgs.Boolean --timeout 300 --req 'data: ["small", "640", "480"]'

    ign service -s /world/simple_demo/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "quadrotor_1", position: {x: 24.90, y:24.85, z:2}'

    ign service -s /mbzirc/target/stream/report --reqtype ignition.msgs.StringMsg_V --reptype ignition.msgs.Boolean --timeout 300 --req 'data: ["large", "640", "480"]'
    ```

1. You can also try reporting wrong image pos and you should see `*_object_id_success_*` events logged. Simulation stops after 3 invalid attempts.

### CODE REVIEW

File: mbzirc_ign/src/GameLogicPlugin.cc [hunk 1]
```diff
@@ -79,6 +89,19 @@ class mbzirc::GameLogicPluginPrivate
...
   };
 
+  /// \brief Information of target in stream
+  public: class TargetInStream
+  {
+    /// \brief Type of target: vessel, small, large
+    public: std::string type;
+
+    /// \brief Image x position
+    public: double x = 0;
```
Comment (caguero) [inline mbzirc_ign/src/GameLogicPlugin.cc:99]: If these are image coordinates, should they be `unsigned int` or similar?

File: mbzirc_ign/src/GameLogicPlugin.cc [hunk 2]
```diff
@@ -310,6 +402,45 @@ class mbzirc::GameLogicPluginPrivate
...
+
+  /// \brief Report of target in stream
+  public: TargetInStream targetInStreamReport;
+
+  /// \brief Current valid target vessel that has been successfully ID'ed.
+  /// This variable is used later when validating reports of target objects.
+  public: std::string currentTargetVessel;
+
+
+  public: std::set<rendering::VisualPtr> targetVesselVisuals;
```
Comment (caguero) [inline mbzirc_ign/src/GameLogicPlugin.cc:433]: Missing docs.

File: mbzirc_ign/src/GameLogicPlugin.cc (hunk 1 above)
Comment (iche033) [reply mbzirc_ign/src/GameLogicPlugin.cc]: yep done. [28876e6](https://github.com/osrf/mbzirc/pull/59/commits/28876e6426e82ad0c5b15c0a7e65b0d2a1e2d1d8)

File: mbzirc_ign/src/GameLogicPlugin.cc (hunk 2 above)
Comment (iche033) [reply mbzirc_ign/src/GameLogicPlugin.cc:433]: added doc. [9c67603](https://github.com/osrf/mbzirc/pull/59/commits/9c6760330ec7a9af7b1b54ef45328a4a45c5ea56)

**Review by caguero — COMMENTED** [review]
I got this error message when running step 7 of your instructions:

```
caguero@cold:~/mbzirc_ws$ ign service -s /world/simple_demo/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "quadrotor_1", position: {x: 24.9.0, y:24.85, z:2}'
[libprotobuf ERROR google/protobuf/text_format.cc:307] Error parsing text-format ignition.msgs.Pose: 1:40: Already saw decimal point or exponent; can't have another one.
[libprotobuf ERROR google/protobuf/text_format.cc:307] Error parsing text-format ignition.msgs.Pose: 1:40: Expected identifier, got: .0
data: true
```

and this is the content of the events file:

```
caguero@cold:~/mbzirc_ws$ tail -f /tmp/ign/mbzirc/logs/events.yml
  total_score: 48
  data: vessel::640::480
- event:
  id: 3
  type: target_reported
  time_sec: 83
  elapsed_real_time: 54
  elapsed_sim_time: 48
  total_score: 48
  data: vessel_id_success
- event:
  id: 4
  type: target_reported_in_stream
  time_sec: 116
  elapsed_real_time: 91
  elapsed_sim_time: 81
  total_score: 80
  data: small::640::480
- event:
  id: 5
  type: target_reported
  time_sec: 116
  elapsed_real_time: 91
  elapsed_sim_time: 81
  total_score: 261
  data: small_object_id_failure_1
- event:
  id: 6
  type: target_reported_in_stream
  time_sec: 128
  elapsed_real_time: 105
  elapsed_sim_time: 93
  total_score: 273
  data: large::640::480
- event:
  id: 7
  type: target_reported
  time_sec: 128
  elapsed_real_time: 105
  elapsed_sim_time: 93
  total_score: 453
  data: large_object_id_failure_1
```

I see a few `object_id_failure`. Is this expected?

**Comment by caguero** [comment]
> I got this error message when running step 7 of your instructions:
> 
> ```
> caguero@cold:~/mbzirc_ws$ ign service -s /world/simple_demo/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "quadrotor_1", position: {x: 24.9.0, y:24.85, z:2}'
> [libprotobuf ERROR google/protobuf/text_format.cc:307] Error parsing text-format ignition.msgs.Pose: 1:40: Already saw decimal point or exponent; can't have another one.
> [libprotobuf ERROR google/protobuf/text_format.cc:307] Error parsing text-format ignition.msgs.Pose: 1:40: Expected identifier, got: .0
> data: true
> ```
> 
> and this is the content of the events file:
> 
> ```
> caguero@cold:~/mbzirc_ws$ tail -f /tmp/ign/mbzirc/logs/events.yml
>   total_score: 48
>   data: vessel::640::480
> - event:
>   id: 3
>   type: target_reported
>   time_sec: 83
>   elapsed_real_time: 54
>   elapsed_sim_time: 48
>   total_score: 48
>   data: vessel_id_success
> - event:
>   id: 4
>   type: target_reported_in_stream
>   time_sec: 116
>   elapsed_real_time: 91
>   elapsed_sim_time: 81
>   total_score: 80
>   data: small::640::480
> - event:
>   id: 5
>   type: target_reported
>   time_sec: 116
>   elapsed_real_time: 91
>   elapsed_sim_time: 81
>   total_score: 261
>   data: small_object_id_failure_1
> - event:
>   id: 6
>   type: target_reported_in_stream
>   time_sec: 128
>   elapsed_real_time: 105
>   elapsed_sim_time: 93
>   total_score: 273
>   data: large::640::480
> - event:
>   id: 7
>   type: target_reported
>   time_sec: 128
>   elapsed_real_time: 105
>   elapsed_sim_time: 93
>   total_score: 453
>   data: large_object_id_failure_1
> ```
> 
> I see a few `object_id_failure`. Is this expected?

Oh, I see, there's a typo. Let me try again.

**Comment by caguero** [comment]
> ```shell
> ```shell
> ign service -s /mbzirc/target/stream/report --reqtype ignition.msgs.StringMsg_V --reptype ignition.msgs.Boolean --timeout 300 --req 'data: ["large", "640", "480"]'
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```

I still see an `data: small_object_id_failure_1` and `large_object_id_failure_1`.

**Comment by caguero** [comment]
> Oh, I see, there's a typo. Let me try again.

Should this line in the instructions

```
ign service -s /world/simple_demo/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "quadrotor_1", position: {x: 24.9.0, y:24.85, z:2}'
```

be

```
ign service -s /world/simple_demo/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "quadrotor_1", position: {x: 24.90, y:24.85, z:2}'
```
?

**Comment by iche033** [comment]
ah thanks for fixing the typos. 

> I still see an data: small_object_id_failure_1 and large_object_id_failure_1.

I think that's because the USV and objects have moved (due to wind and waves). Could you try opening `Image Display` gui plugin and see if they objects are in the center of the camera view before invoking the service call? If not, you'll need to tweak the UAV position so that the objects are positioned in the center of the image (or specify different image coordinates which is harder to do).

**Comment by caguero** [comment]
I confirm that I get a successful confirmation when the objects are centered in the image.

```
- event:
  id: 4
  type: target_reported_in_stream
  time_sec: 46
  elapsed_real_time: 35
  elapsed_sim_time: 30
  total_score: 30
  data: small::640::480
- event:
  id: 5
  type: target_reported
  time_sec: 46
  elapsed_real_time: 35
  elapsed_sim_time: 30
  total_score: 30
  data: small_object_id_success
- event:
  id: 6
  type: target_reported_in_stream
  time_sec: 56
  elapsed_real_time: 46
  elapsed_sim_time: 40
  total_score: 39
  data: large::640::480
- event:
  id: 7
  type: target_reported
  time_sec: 56
  elapsed_real_time: 46
  elapsed_sim_time: 40
  total_score: 39
  data: large_object_id_success

```

I changed the sea state and wind by running:

```
 ./set_sea_state.sh 0 --wind-dir 0 0 0
```



> I think that's because the USV and objects have moved (due to wind and waves). Could you try opening `Image Display` gui plugin and see if they objects are in the center of the camera view before invoking the service call? If not, you'll need to tweak the UAV position so that the objects are positioned in the center of the image (or specify different image coordinates which is harder to do).

**Review by caguero — APPROVED** [review]
